File: functions/filtering.py
import numpy as np
import cv2 as cv
from skimage.filters import unsharp_mask

#Smoothing images/ Low Pass Filtering - OpenCV ready
#def blur(img_src, ksizex:int = 5, ksizey:int = 5) -> 	dst
#def GaussianBlur(img_src, ksizex:int = 5, ksizey:int = 5) -> 	dst
#ksize positive and odd 3,5,7
#	cv.GaussianBlur(	src, ksize, sigmaX=0)
#sigmaX = 0
#def medianBlur(img_src, ksize: int = 5	) -> 	dst
#def bilateralFilter(img_src, d: int = 5 , sigmaColor: int = 50, sigmaSpace: int = 50) -> 	dst
#sigma > 5, d filter size < 10
#def filter2D(img_src, ddepth: int = -1, img_kernel) -> 	dst
#need a funct that creates a kernel, maybe

#Image gradients / High Pass Filtering
def gradient(img, filterType: str = "Laplacian", dx: int = 1, dy: int = 1,
		ksize: int = 3, returnFloats: bool = False):
	#param: filterType: Sobel, Scharr and Laplacian
	#default values, see documentation
	#ksize must be positive and odd: must be exactly 1,3,5,7

	if filterType == "Laplacian":
		imgx64F = cv.Laplacian(img, cv.CV_64F, ksize = ksize)
	elif filterType == "Sobel":
		imgx64F = cv.Sobel(img, cv.CV_64F, dx, dy, ksize = ksize)
	elif filterType == "Scharr":
		imgx64F = cv.Scharr(img, cv.CV_64F, dx, dy) # den exei ksize
	else:
		raise Exception('In gradient, filter type is not supported')
	
	if returnFloats:
		return {'floatMatr': imgx64F}

	abs_img64f = np.absolute(imgx64F)
	img_8u = np.uint8(abs_img64f)

	return {'img': img_8u}

#More filtering
#Unsharp masking
def unsharpMask(img, radius: float = 1, amount: float = 1):
	isGray = (img.ndim == 2)
	unsharp_img = unsharp_mask(image = img, radius=radius, amount=amount, preserve_range=True, channel_axis=(None if isGray else 2))

	return {'img': unsharp_img}

#Edge detection
#Canny ED - OpenCV ready
#def Canny(img_image, threshold1: int = 100, threshold2: int = 200, apertureSize: int = 3, L2gradient: bool = false) -> 	edges
# 0 512 thresh diasthma
# sobel kernel 1,3,5,7

File: myNodes/filtering.js
'use strict'

module.exports = function(RED) {

  //blur
  function Blur(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let request = {
        funcName: "blur",
        params: {
          img_src: msg.payload,
          ksize: [parseInt(config.ksizex), parseInt(config.ksizey)]
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("blur", Blur);


  //GaussianBlur
  function GaussianBlur(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let request = {
        funcName: "GaussianBlur",
        params: {
          img_src: msg.payload,
          ksize: [parseInt(config.ksizex), parseInt(config.ksizey)],
          sigmaX: 0
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("gaussian blur", GaussianBlur);

  //medianBlur
  function MedianBlur(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let request = {
        funcName: "medianBlur",
        params: {
          img_src: msg.payload,
          ksize: parseInt(config.ksize)
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("median blur", MedianBlur);

  //bilateralFilter
  function BilateralFilter(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let request = {
        funcName: "bilateralFilter",
        params: {
          img_src: msg.payload,
          d: parseInt(config.dd),
          sigmaColor: parseInt(config.sigmaColor),
          sigmaSpace: parseInt(config.sigmaSpace)
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("bilateral filter", BilateralFilter);

  //filter2D
  //exclude it for now, because we dont have the previous node that creates the 2d filter
  /*function GeneralFilter(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let request = {
        funcName: "filter2D",
        params: {
          img_src: msg.payload.img,//TODO chec etsi legetai?
          ddepth: -1,
          kernel: msg.payload.kernel//TODO ayto thelei ftiaksimo to prin tou
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("general filter", GeneralFilter);*/

  //gradient
  function Gradient(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let dx = 0, dy = 0;
      if (config.filterType == "Scharr") {
        dx = parseInt(config.scharrDerivative);
        dy = 1 - dx;
      } else {
        dx = parseInt(config.dx);
        dy = parseInt(config.dy);
      }

      let request = {
        funcName: "gradient",
        params: {
          img: msg.payload,
          filterType: config.filterType,
          dx: dx,
          dy: dy,
          ksize: parseInt(config.ksize)
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("gradient", Gradient);

  //unsharpMask
  function UnsharpMasking(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let request = {
        funcName: "unsharpMask",
        params: {
          img: msg.payload,
          radius: parseFloat(config.radius),
          amount: parseFloat(config.amount)
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("unsharp masking", UnsharpMasking);


  //Canny
  function CannyED(config) {
    RED.nodes.createNode(this, config);
    let node = this;

    let broker = RED.nodes.getNode(config.broker);

    node.on('input', function(msg, nodeSend, nodeDone) {

      let request = {
        funcName: "Canny",
        params: {
          img_image: msg.payload,
          threshold1: parseInt(config.threshold1),
          threshold2: parseInt(config.threshold2),
          apertureSize: parseInt(config.apertureSize),
          L2gradient: config.L2gradient
        }
      }
      broker.sendMsg(request, node, msg.times, function callback(response){

        msg.payload = response.data.results.img;
        msg.key = 'img';
        nodeSend(msg);
        nodeDone();
      });
    });
  }
  RED.nodes.registerType("canny edge detection", CannyED);
}
